### PR TITLE
Adjust Vec to build on stable Rust

### LINKED ISSUE #5: Missing bounds check in later `Vec` allocation code
https://github.com/rust-lang/rust/issues/31941

### LINKED ISSUE #70: Implementing Vec final code does not compile
The final example does not compile (just clicking the run button), with following messages:
```
   Compiling playground v0.0.1 (file:///playground)
warning: use of deprecated item 'std::heap::Heap': type renamed to `Global`
 --> src/main.rs:9:32
  |
9 | use std::heap::{Alloc, Layout, Heap};
  |                                ^^^^
  |
  = note: #[warn(deprecated)] on by default

warning: use of deprecated item 'std::heap::Heap': type renamed to `Global`
  --> src/main.rs:34:27
   |
34 |                 let ptr = Heap.alloc(Layout::array::<T>(1).unwrap());
   |                           ^^^^

warning: use of deprecated item 'std::heap::Heap': type renamed to `Global`
  --> src/main.rs:38:27
   |
38 |                 let ptr = Heap.realloc(self.ptr.as_ptr() as *mut _,
   |                           ^^^^

warning: use of deprecated item 'std::heap::Heap': type renamed to `Global`
  --> src/main.rs:47:29
   |
47 |                 Err(err) => Heap.oom(err),
   |                             ^^^^

warning: use of deprecated item 'std::heap::Heap': type renamed to `Global`
  --> src/main.rs:61:17
   |
61 |                 Heap.dealloc(self.ptr.as_ptr() as *mut _,
   |                 ^^^^

error[E0308]: mismatched types
  --> src/main.rs:38:40
   |
38 |                 let ptr = Heap.realloc(self.ptr.as_ptr() as *mut _,
   |                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `std::ptr::NonNull`, found *-ptr
   |
   = note: expected type `std::ptr::NonNull<std::heap::Opaque>`
              found type `*mut _`

error[E0308]: mismatched types
  --> src/main.rs:40:40
   |
40 |                                        Layout::array::<T>(new_cap).unwrap());
   |                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected usize, found struct `std::heap::Layout`
   |
   = note: expected type `usize`
              found type `std::heap::Layout`

error[E0599]: no method named `oom` found for type `std::heap::Global` in the current scope
  --> src/main.rs:47:34
   |
47 |                 Err(err) => Heap.oom(err),
   |                                  ^^^

error[E0308]: mismatched types
  --> src/main.rs:61:30
   |
61 |                 Heap.dealloc(self.ptr.as_ptr() as *mut _,
   |                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `std::ptr::NonNull`, found *-ptr
   |
   = note: expected type `std::ptr::NonNull<std::heap::Opaque>`
              found type `*mut _`

error: aborting due to 4 previous errors

Some errors occurred: E0308, E0599.
For more information about an error, try `rustc --explain E0308`.
error: Could not compile `playground`.

To learn more, run the command again with --verbose.
```

Also previous examples during Implementing Vec chapter are using even older allocator API then this final code.

### LINKED ISSUE #83: inconsistent, outdated Vec chapter
#73 switched to a newer version of the alloc API, but not the final stabilized version after the changes in https://github.com/rust-lang/rust/pull/51241. It also didn't update any of the earlier code in the chapter (which still uses `heap::reallocate`, etc), so now the chapter is inconsistent and confusing to the reader when `vec-final.md` isn't composed of all the code they've been shown step-by-step so far.

A lot of the explanatory text also seems dated:
> Because all the best tools for writing unsafe code are unstable, this project will only work on nightly (as of Rust 1.9.0). With the exception of the allocator API, much of the unstable code we'll use is expected to be stabilized in a similar form as it is today.

>Unfortunately the mechanism for stating that your value is non-zero is unstable and unlikely to be stabilized soon. As such we're just going to take the hit and use std's Unique:

### LINKED ISSUE #160: Rewrite "Implementing Vec" chapter - Unique is permanently unstable
This chapter use `Unique` to demonstrate how to implement a Vec. But it is
permanently unstable and can only be used by std. Also in
https://github.com/rust-lang/rust/blob/478464570e60523adc6d303577d1782229ca1f93/src/libcore/ptr/unique.rs#L28-L34
it recommends that we should use `NonNull` instead and consider `PhantomData<T>`.

### PR DESCRIPTION
**Pull request by blkerby**
Hi, first time contributing to the Nomicon here. I'm aware this is a somewhat large proposed change (larger than I anticipated it being when I began) but I hope it will be useful.

The chapter on implementing Vec was written to build on nightly Rust. From the discussion in 9.1 (Layout), the primary motivation for this was to get an optimized non-nullable pointer type. Since `std::ptr::NonNull` is now stable, this can achieved on stable Rust. This pull request makes the changes throughout the chapter necessary to implement Vec using `std::ptr::NonNull` instead of the unstable `Unique`.

Currently much of the code in the main body of the chapter does not compile on the latest nightly Rust because it relies on unstable APIs which have changed (for instance, `std::alloc::oom`, `alloc::heap::{allocate, reallocate, deallocate}` are gone, and `Unique` no longer implements `Deref`). The final code in the last section of the chapter compiles, because it has been updated to use current unstable APIs such as `std::alloc::Global`. This pull request updates the main body of the chapter as well as the final code so that they are consistent with each other, both now using the stable allocation functions in `std::alloc` (while mentioning the new unstable type `std::alloc::Global` in the text).

Instead of incorporating `NonNull` directly into `Vec`, `RawVec`, etc. as this pull request does, one alternative would be to implement a stable version of `Unique` using `NonNull` internally, and then include this `Unique` in `Vec`, `RawVec`, etc. I briefly started going down that pathway before concluding that for this to be convenient quite a few methods would need to be implemented for this new `Unique`, which would seem to create a distraction from the main goal and add an unnecessary layer of abstraction. But this would be another option if there's a preference for going that way.

Fixes #5, fixes #70, fixes #83, fixes #160

### CODE REVIEW

**Comment by blkerby** [comment]
FYI, after finishing the changes, I made a pass through the chapter and tested the cumulative code at each step, confirming that it compiles successfully on stable Rust.

**Comment by blkerby** [comment]
Fixes #160

**Comment by blkerby** [comment]
Fixes #83

**Comment by blkerby** [comment]
Fixes #70. The run button now works on the final code.

**Comment by blkerby** [comment]
Fixes #5

**Comment by skade** [comment]
I want to register interest in getting this PR rebased and merged.

**Comment by blkerby** [comment]
I wouldn't mind rebasing and making the needed changes to get this ready again; could we get a reviewer/assignee first so we can know if the maintainers are interested in getting it merged? @JohnTitor, would you be interested in reviewing, or helping us find someone?

**Comment by JohnTitor** [comment]
Hey sorry for the delay! I'll try to find some time to review in a few days and it'd be great if you could rebase meanwhile.

**Comment by blkerby** [comment]
@JohnTitor I resolved the conflicts and rebased, so it should be ready for review now!

File: src/vec-alloc.md
Comment (JohnTitor) [inline src/vec-alloc.md:176]:
```suggestion
            // `Layout::array` checks that the number of bytes is <= usize::MAX,
```

File: src/vec-dealloc.md
Comment (JohnTitor) [inline src/vec-dealloc.md:20]:
```suggestion
            unsafe {
                alloc::dealloc(self.ptr.as_ptr() as *mut u8, layout);
```

File: src/vec-into-iter.md
Comment (JohnTitor) [inline src/vec-into-iter.md:141]:
```suggestion
            unsafe {
                alloc::dealloc(self.buf.as_ptr() as *mut u8, layout);
```

File: src/vec-raw.md
Comment (JohnTitor) [inline src/vec-raw.md:80]:
```suggestion
```

File: src/vec-zsts.md
Comment (JohnTitor) [inline src/vec-zsts.md:41]:
```suggestion
        // `NonNull::dangling()` doubles as "unallocated" and "zero-sized allocation"
```

File: src/vec-zsts.md
Comment (JohnTitor) [inline src/vec-zsts.md:60]:
```suggestion
            // `Layout::array` checks that the number of bytes is <= usize::MAX,
```

File: src/vec-final.md
Comment (JohnTitor) [inline src/vec-final.md:251]:
```suggestion
        let len = (self.end as usize - self.start as usize) /
```

File: src/vec-final.md
Comment (JohnTitor) [inline src/vec-final.md:24]:
```suggestion
        // `NonNull::dangling()` doubles as "unallocated" and "zero-sized allocation"
```

File: src/vec-final.md
Comment (JohnTitor) [inline src/vec-final.md:43]:
```suggestion
            // `Layout::array` checks that the number of bytes is <= usize::MAX,
```

File: src/vec-final.md
Comment (JohnTitor) [inline src/vec-final.md:339]:
```suggestion
#
```

File: src/vec-final.md
Comment (JohnTitor) [inline src/vec-final.md:357]:
```suggestion
#
```

File: src/vec-final.md
Comment (JohnTitor) [inline src/vec-final.md:370]:
```suggestion
#
```

File: src/vec-final.md
Comment (JohnTitor) [inline src/vec-final.md:387]:
```suggestion
#
```

File: src/vec-final.md
Comment (JohnTitor) [inline src/vec-final.md:393]:
```suggestion
#
```

File: src/vec-final.md
Comment (JohnTitor) [inline src/vec-final.md:395]:
```suggestion
#
```

File: src/vec-final.md
Comment (JohnTitor) [inline src/vec-final.md:399]:
```suggestion
#
```

**Review by JohnTitor — COMMENTED** [review]
Sorry for the delay! Just read and the implementation itself looks good to me, just left suggestions for the style issues.

**Comment by blkerby** [comment]
Thanks, I applied all the suggested changes.

**Review by JohnTitor — APPROVED** [review]
LGTM! But I'm going to ask the libs team to review since this is a major change and I'm not sure if I've got the whole picture correctly (related Zulip topic: https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/review.20request.20for.20the.20nomicon.20PR), I'll wait for a week or so to get some more eyes. Sorry to keep you waiting.

File: src/vec-dealloc.md
Comment (Amanieu) [inline src/vec-dealloc.md:8]: Note that the above is only true if LLVM can prove that the loop is finite. This is true in the case of `Vec`, but for example LLVM has no way to know that a linked list has no cycles.

File: src/vec-final.md
Comment (Amanieu) [inline src/vec-final.md:114]: All the calls to `.offset` can be replaced with `.add` which avoids the need to cast to `isize`.

File: src/vec-insert-remove.md
Comment (Amanieu) [inline src/vec-insert-remove.md:26]: Formatting.

File: src/vec-alloc.md
Comment (jyn514) [inline src/vec-alloc.md:45]: At some point I really want to implement https://github.com/rust-lang/rust/issues/77974.

**Comment by JohnTitor** [comment]
It's been two weeks, @blkerby could you address Amanieu's reviews? Once it's done, I'm going to merge this finally.

File: src/vec-dealloc.md
Comment (blkerby) [reply src/vec-dealloc.md:8]: Makes sense. And based on the docs for [needs_drop](https://doc.rust-lang.org/std/mem/fn.needs_drop.html) it sounds like even in the `Vec` case LLVM might not eliminate the loop in debug builds. In any case, it is probably fragile to rely on LLVM for this behavior. I'm thinking of a few options for how we could address this:

1. Leave the code as is, but change the text to make the risk more clear.
2. Use a `mem::needs_drop` to skip the loop if `T` is `!Drop`.
3. Instead of an explicit loop, use `ptr::drop_in_place` (which is what the standard library `Vec` does).
4. Start with the current code, but within the section discuss its pitfalls and iterate on it, progressing to 2 and/or 3.

I would lean toward option 4, since currently this section is very short and could benefit from more discussion and examples to illustrate the usage of `mem::needs_drop` and `ptr::drop_in_place`. If we go with 2, 3, or 4, it will also require changes to `vec-raw.md` and `vec-final.md` for consistency. If you don't mind, do you think I could create a separate PR for this change, since it addresses a separate issue from the current PR?

File: src/vec-final.md
Comment (blkerby) [reply src/vec-final.md:114]: Nice! There are a couple dozen or so places throughout the chapter where this change could be made. Would you mind if I create a separate PR for this?

File: src/vec-insert-remove.md
Comment (blkerby) [reply src/vec-insert-remove.md]: Fixed, thanks.

Comment (blkerby) [reply src/vec-insert-remove.md]: Fixed, thanks.

**Comment by blkerby** [comment]
@JohnTitor I think the reviews are identifying additional ways to improve the Vec chapter that are orthogonal to the changes in this PR. If you think it makes sense, could we merge this one and then tackle those in separate PRs?

**Comment by JohnTitor** [comment]
> I think the reviews are identifying additional ways to improve the Vec chapter that are orthogonal to the changes in this PR. If you think it makes sense, could we merge this one and then tackle those in separate PRs?

Works for me, I'd also open another PR. Thanks for your awesome work, @blkerby!

File: src/vec-dealloc.md
Comment (Amanieu) [reply src/vec-dealloc.md:8]: I agree that option 4 sounds like the best way to go. It's fine to do this in a separate PR.

**Comment by Gankra** [comment]
hey, thanks a lot!!

**Pull request merged**
